### PR TITLE
allow use of a byref parameter inside of nameof

### PR DESCRIPTION
**Pull request by TyOverby**
**Customer scenario**
The customer uses a byref parameter inside of a nameof that is inside of a closure.

**Bugs this fixes:** 

https://github.com/dotnet/roslyn/issues/10839

**Workarounds, if any**

Don't use nameof, use strings instead.

**Risk**

Low.

**Performance impact**

No added allocations.  Additional code is extremely minimal.

**Is this a regression from a previous update?**

No

**How was the bug found?**

Reported by a customer in https://github.com/dotnet/roslyn/issues/10467

### CODE REVIEW

**Comment by TyOverby** [comment]
@dotnet/roslyn-compiler Review please.

File: src/Compilers/VisualBasic/Portable/Lowering/Diagnostics/DiagnosticsPass.vb [hunk 1]
```diff
@@ -71,7 +72,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                     ' is a QueryLambda, before we reach parameter's container. 
                     If Binder.IsTopMostEnclosingLambdaAQueryLambda(_containingSymbol, parameterSymbolContainingSymbol) Then
                         Binder.ReportDiagnostic(Me._diagnostics, node.Syntax, ERRID.ERR_CannotLiftByRefParamQuery1, parameterSymbol.Name)
-                    Else
+                    ElseIf Not Me._insideNameof Then
```
Comment (AdamSpeight2008) [inline src/Compilers/VisualBasic/Portable/Lowering/Diagnostics/DiagnosticsPass.vb]: Maybe possible to remove the `Me.` prefix.

**Comment by TyOverby** [comment]
I've decided to not push this for dev15, instead retargeting the post-dev15 branch.

File: src/Compilers/VisualBasic/Test/Semantic/Semantics/NameOfTests.vb [hunk 2]
```diff
@@ -3544,5 +3544,28 @@ TP
...
+    End Sub
+    Sub Main()
+        Dim x =  5
+        DoSomething(x)
+    End Sub
+End Module
+                </file>
+                </compilation>
+            Dim comp = CreateCompilationWithMscorlibAndVBRuntime(compilationDef, TestOptions.DebugExe)
+            CompileAndVerify(comp, expectedOutput:="x").VerifyDiagnostics()
```
Comment (jaredpar) [inline src/Compilers/VisualBasic/Test/Semantic/Semantics/NameOfTests.vb]: Please add a test that uses nameof inside a LINQ clause.

File: src/Compilers/VisualBasic/Portable/Lowering/Diagnostics/DiagnosticsPass.vb [hunk 3]
```diff
@@ -67,12 +67,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                 ' cannot access ByRef parameters in Lambdas
                 Dim parameterSymbolContainingSymbol As Symbol = parameterSymbol.ContainingSymbol
 
-                If _containingSymbol IsNot parameterSymbolContainingSymbol Then
+                If _containingSymbol IsNot parameterSymbolContainingSymbol And Not _insideNameof Then
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Lowering/Diagnostics/DiagnosticsPass.vb]: `AndAlso`

Comment (TyOverby) [inline src/Compilers/VisualBasic/Portable/Lowering/Diagnostics/DiagnosticsPass.vb]: Done

File: src/Compilers/VisualBasic/Test/Semantic/Semantics/NameOfTests.vb [hunk 4]
```diff
@@ -3544,5 +3544,51 @@ TP
...
+        Public Sub NameOfByRefInQuery()
+            Dim compilationDef =
+                <compilation>
+                    <file name="a.vb">
+Imports System.Linq
+
+Module Program
+    Sub DoSomething(ByRef x As Integer)
+        Dim f = from x in {1, 2, 3}
+                select nameof(x)
```
Comment (cston) [inline src/Compilers/VisualBasic/Test/Semantic/Semantics/NameOfTests.vb]: Is `x` the parameter or a separate iteration variable? Consider changing iteration variable name so it's clear.

**Comment by cston** [comment]
LGTM

**Comment by TyOverby** [comment]
@dotnet-bot test this please

**Pull request merged**